File: node/src/controllers/healthcheck.controller.ts
import express, { Request, Response } from 'express'
import { Exception, IHealthCheckService, Result } from 'healthz-curler-shared-js'
import { Ok } from 'healthz-curler-shared-js'

const getHealthCheckService = ({ container }: Request) => container.get<IHealthCheckService>('IHealthCheckService')

const getAll = async (req: Request, res: Response) => {
  const healthCheckService = getHealthCheckService(req)
  const result = await healthCheckService.getAll()
  if (result.ok) return res.status(200).json(result.value)
  return res.status(400).json(result.error)
}

const getById = async (req: Request, res: Response) => {
  const { id } = req.params
  const healthCheckService = getHealthCheckService(req)
  const validationResult = validateRequest(id)
  if (!validationResult.ok) return res.status(400).json(validationResult.error)

  const result = await healthCheckService.get(id!)
  if (result.ok) return res.status(200).json(result.value)
  return res.status(500).json(result.error)
}

const create = async (req: Request, res: Response) => {
  const healthCheckService = getHealthCheckService(req)
  const result = await healthCheckService.create(req.body)
  if (result.ok) return res.status(200).json(result.value)
  return res.status(500).json(result.error)
}

const updateById = async (req: Request, res: Response) => {
  const { id } = req.params
  const healthCheckService = getHealthCheckService(req)
  const validationResult = validateRequest(id)
  if (!validationResult.ok) return res.status(400).json(validationResult.error)

  const result = await healthCheckService.update(id!, req.body)
  if (result.ok) return res.status(200).json(result.value)
  return res.status(500).json(result.error)
}

const deleteById = async (req: Request, res: Response) => {
  const { id } = req.params
  const healthCheckService = getHealthCheckService(req)
  const validationResult = validateRequest(id)
  if (!validationResult.ok) return res.status(400).json(validationResult.error)

  const result = await healthCheckService.delete(id!)
  if (result.ok) return res.status(200).json(result.value)
  return res.status(500).json(result.error)
}

const validateRequest = (id: string | undefined): Result<boolean, Error> => {
  if (!id) return new Exception('Id is required')
  return Ok(true)
}

const healthCheckRouter = express.Router()
healthCheckRouter.get('/', getAll)
healthCheckRouter.post('/', create)
healthCheckRouter.get('/:id', getById)
healthCheckRouter.put('/:id', updateById)
healthCheckRouter.delete('/:id', deleteById)

export { healthCheckRouter }

File: go/cmd/server/controllers/healthcheck_controller.go
package controllers

import (
	"errors"

	"github.com/gofiber/fiber/v2"
	"github.com/marlpativ/healthz-curler/cmd/server/container"
	"github.com/marlpativ/healthz-curler/internal/models"
	"github.com/marlpativ/healthz-curler/internal/services"
)

func getHealthCheckService(c *fiber.Ctx) services.HealthCheckService {
	// Get container from context
	ctr := c.Locals("container")
	return ctr.(*container.Container).HealthCheckService
}

func GetAllHealthChecks(c *fiber.Ctx) error {
	healthCheckService := getHealthCheckService(c)
	healthChecks, err := healthCheckService.GetAll()
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": err.Error(),
		})
	}
	return c.Status(fiber.StatusOK).JSON(healthChecks)
}

func validateHealthCheckRequest(c *fiber.Ctx) error {
	id := c.Params("id")
	if id == "" {
		return errors.New("id is required")
	}
	return nil
}

func GetHealthCheck(c *fiber.Ctx) error {
	if err := validateHealthCheckRequest(c); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": err.Error()})
	}
	healthCheckService := getHealthCheckService(c)
	id := c.Params("id")
	healthCheck, err := healthCheckService.Get(id)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"error": err.Error()})
	}
	return c.Status(fiber.StatusOK).JSON(healthCheck)
}

func createHealthCheck(c *fiber.Ctx) error {
	healthCheckService := getHealthCheckService(c)
	healthCheck := new(models.HealthCheck)
	if err := c.BodyParser(healthCheck); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": err.Error()})
	}
	res, err := healthCheckService.Create(*healthCheck)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"error": err.Error()})
	}
	return c.Status(fiber.StatusCreated).JSON(res)
}

func updateHealthCheck(c *fiber.Ctx) error {
	if err := validateHealthCheckRequest(c); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": err.Error()})
	}
	healthCheckService := getHealthCheckService(c)
	healthCheck := new(models.HealthCheck)
	if err := c.BodyParser(healthCheck); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": err.Error()})
	}
	id := c.Params("id")
	res, err := healthCheckService.Update(id, *healthCheck)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"error": err.Error()})
	}
	return c.Status(fiber.StatusCreated).JSON(res)
}

func deleteHealthCheck(c *fiber.Ctx) error {
	if err := validateHealthCheckRequest(c); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": err.Error()})
	}
	healthCheckService := getHealthCheckService(c)
	id := c.Params("id")
	res, err := healthCheckService.Delete(id)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"error": err.Error()})
	}
	return c.Status(fiber.StatusCreated).JSON(res)
}

func SetupHealthCheckRoute(app fiber.Router) {
	healthCheckRouter := app.Group("/healthcheck")
	healthCheckRouter.Get("/:id", GetAllHealthChecks)
	healthCheckRouter.Put("/:id", updateHealthCheck)
	healthCheckRouter.Delete("/:id", deleteHealthCheck)
	healthCheckRouter.Post("/", createHealthCheck)
	healthCheckRouter.Get("/", GetAllHealthChecks)
}
